File: benchmark.m
% ZeroMQ Frame Size Benchmark for MATLAB
% Requires https://github.com/fagg/matlab-zmq

function benchmark(type)
	if strcmpi(type,'server')
		run_zmq_server(5555)
	else
		benchmark_zmq_frame_sizes()
	end
end

function results = benchmark_zmq_frame_sizes()
    % Configuration parameters
    port = 5555;
    header_size = 100;
    data_size = 500000;  % 500 KB
    num_runs = 5;
    
    % Test various chunk sizes
    chunk_sizes = [
        1024,       % 1KB
        8192,       % 8KB
        16384,      % 16KB
        32768,      % 32KB
        65536,      % 64KB
        131072,     % 128KB
        262144,     % 256KB
        524288      % 512KB (single frame for data)
    ];
    
    % Initialize results structure
    results = struct('chunk_size', {}, 'num_frames', {}, 'latency_ms', {}, 'throughput_mbps', {});
    
    % Create ZeroMQ context
    context = zmq.core.ctx_new();
    
    % Start server in a separate MATLAB instance
    disp('Starting server...');
    disp('To run the server, start another MATLAB instance and run:');
    disp(['run_zmq_server(' num2str(port) ');']);
    input('Press Enter when the server is running...');
    
    % Generate test data
    header = uint8(repmat('H', 1, header_size));
    data = uint8(repmat('D', 1, data_size));
    
    fprintf('\nStarting benchmark with %d byte header and %.1f KB data\n', header_size, data_size/1000);
    fprintf('Running %d tests per chunk size\n\n', num_runs);
    
    % Run benchmarks for each chunk size
    for i = 1:length(chunk_sizes)
        chunk_size = chunk_sizes(i);
        fprintf('Benchmarking chunk size: %d bytes\n', chunk_size);
        
        run_latencies = zeros(1, num_runs);
        run_throughputs = zeros(1, num_runs);
        
        % Split data into chunks and prepare frames
        chunks = {};
        for j = 1:chunk_size:length(data)
            end_idx = min(j + chunk_size - 1, length(data));
            chunks{end+1} = data(j:end_idx);
        end
        
        num_frames = length(chunks) + 1;  % +1 for header
        
        % Calculate total bytes to be sent
        total_bytes = header_size;
        for j = 1:length(chunks)
            total_bytes = total_bytes + length(chunks{j});
        end
        
        % Run multiple tests for this chunk size
        for run = 1:num_runs
            % Create and connect socket for this run
            socket = zmq.core.socket(context, 'ZMQ_REQ');
            zmq.core.connect(socket, ['tcp://localhost:' num2str(port)]);
            
            % Warm-up round
            send_multipart_message(socket, header, chunks);
            receive_multipart_message(socket);
            
            % Test round
            tic;
            send_multipart_message(socket, header, chunks);
            receive_multipart_message(socket);
            elapsed_time = toc;
            
            % Calculate metrics
            latency_ms = elapsed_time * 1000;
            throughput_mbps = (total_bytes * 8 / 1000000) / elapsed_time;
            
            run_latencies(run) = latency_ms;
            run_throughputs(run) = throughput_mbps;
            
            fprintf('  Run %d: Latency=%.2fms, Throughput=%.2fMbps\n', ...
                run, latency_ms, throughput_mbps);
            
            % Close socket
            zmq.core.close(socket);
            pause(0.5);  % Brief pause between runs
        end
        
        % Store average results
        result = struct(...
            'chunk_size', chunk_size, ...
            'num_frames', num_frames, ...
            'latency_ms', mean(run_latencies), ...
            'throughput_mbps', mean(run_throughputs));
        
        results(i) = result;
    end
    
    % Clean up
    zmq.core.ctx_shutdown(context);
    zmq.core.ctx_term(context);
    
    % Print results table
    disp('=== BENCHMARK RESULTS ===');
    fprintf('%-20s %-10s %-20s %-20s\n', 'Chunk Size (bytes)', '# Frames', 'Avg Latency (ms)', 'Avg Throughput (Mbps)');
    disp(repmat('-', 1, 70));
    
    for i = 1:length(results)
        fprintf('%-20d %-10d %-20.2f %-20.2f\n', ...
            results(i).chunk_size, ...
            results(i).num_frames, ...
            results(i).latency_ms, ...
            results(i).throughput_mbps);
    end
    
    % Find best results
    [~, min_latency_idx] = min([results.latency_ms]);
    [~, max_throughput_idx] = max([results.throughput_mbps]);
    
    disp('=== SUMMARY ===');
    fprintf('Lowest latency: %d bytes (%.2fms)\n', ...
        results(min_latency_idx).chunk_size, results(min_latency_idx).latency_ms);
    fprintf('Highest throughput: %d bytes (%.2fMbps)\n', ...
        results(max_throughput_idx).chunk_size, results(max_throughput_idx).throughput_mbps);
    
    % Plot results
    figure;
    
    subplot(2, 1, 1);
    plot([results.chunk_size], [results.latency_ms], 'o-', 'LineWidth', 2);
    set(gca, 'XScale', 'log');
    xlabel('Chunk Size (bytes)');
    ylabel('Latency (ms)');
    title('ZeroMQ Frame Size vs. Latency');
    grid on;
    
    subplot(2, 1, 2);
    plot([results.chunk_size], [results.throughput_mbps], 'o-', 'LineWidth', 2);
    set(gca, 'XScale', 'log');
    xlabel('Chunk Size (bytes)');
    ylabel('Throughput (Mbps)');
    title('ZeroMQ Frame Size vs. Throughput');
    grid on;
    
    sgtitle('ZeroMQ Frame Size Benchmark Results');
end

% Helper function to send multipart message
function send_multipart_message(socket, header, chunks)
    % Send header with more flag
    zmq.core.send(socket, header, 'ZMQ_SNDMORE');
    
    % Send all data chunks except the last one with more flag
    for i = 1:length(chunks)-1
        zmq.core.send(socket, chunks{i}, 'ZMQ_SNDMORE');
    end
    
    % Send the last chunk with no more flag
    if ~isempty(chunks)
        zmq.core.send(socket, chunks{end});
    else
        % If there are no chunks, send empty message to complete the send
        zmq.core.send(socket, uint8([]));
    end
end

% Helper function to receive multipart message
function frames = receive_multipart_message(socket)
    frames = {};
    
    % Receive first part
    [frames{1}, more] = zmq.core.recv(socket);
    
    % Continue receiving parts while more flag is set
    while more
        [frames{end+1}, more] = zmq.core.recv(socket);
    end
end

% Server function to be run in a separate MATLAB instance
function run_zmq_server(port)
    disp(['Starting ZeroMQ server on port ' num2str(port) '...']);
    
    % Create context and socket
    context = zmq.core.ctx_new();
    socket = zmq.core.socket(context, 'ZMQ_REP');
    zmq.core.bind(socket, ['tcp://*:' num2str(port)]);
    
    disp('Server running. Press Ctrl+C to stop.');
    
    try
        while true
            % Receive multipart message
            frames = {};
            [frames{1}, more] = zmq.core.recv(socket);
            
            % Continue receiving parts while more flag is set
            while more
                [frames{end+1}, more] = zmq.core.recv(socket);
            end
            
            fprintf('Received %d frames\n', length(frames));
            
            % Echo back the message
            for i = 1:length(frames)-1
                zmq.core.send(socket, frames{i}, 'ZMQ_SNDMORE');
            end
            zmq.core.send(socket, frames{end});
        end
    catch e
        disp('Server stopped.');
        disp(e.message);
    end
    
    % Clean up
    zmq.core.close(socket);
    zmq.core.ctx_shutdown(context);
    zmq.core.ctx_term(context);
end

File: benchmark.py
import zmq
import time
import threading
import statistics
import argparse
from dataclasses import dataclass
from typing import List, Tuple

@dataclass
class BenchmarkResult:
    chunk_size: int
    num_frames: int
    latency_ms: float
    throughput_mbps: float


def start_server(port: int) -> threading.Thread:
    """Starts a ZeroMQ server in a separate thread that echoes back received messages."""
    def _server_thread():
        context = zmq.Context()
        socket = context.socket(zmq.REP)
        socket.bind(f"tcp://*:{port}")
        
        print("Server started and listening...")
        try:
            while True:
                frames = socket.recv_multipart()
                socket.send_multipart(frames)  # Echo back the same frames
        except zmq.error.ContextTerminated:
            print("Server shutting down...")
    
    thread = threading.Thread(target=_server_thread, daemon=True)
    thread.start()
    return thread


def run_client_test(port: int, header: bytes, data: bytes, chunk_size: int) -> BenchmarkResult:
    """Runs a client test with the specified frame sizes and returns performance metrics."""
    context = zmq.Context()
    socket = context.socket(zmq.REQ)
    socket.connect(f"tcp://localhost:{port}")
    
    # Split the data into chunks of the specified size
    chunks = []
    for i in range(0, len(data), chunk_size):
        chunks.append(data[i:i + chunk_size])
    
    # Create frames
    frames = [header] + chunks
    num_frames = len(frames)
    total_bytes = sum(len(frame) for frame in frames)
    
    # Warm-up round
    socket.send_multipart(frames)
    socket.recv_multipart()
    
    # Test round
    start_time = time.time()
    socket.send_multipart(frames)
    socket.recv_multipart()
    end_time = time.time()
    
    elapsed_time = end_time - start_time
    latency_ms = elapsed_time * 1000
    throughput_mbps = (total_bytes * 8 / 1000000) / elapsed_time
    
    socket.close()
    
    return BenchmarkResult(
        chunk_size=chunk_size,
        num_frames=num_frames,
        latency_ms=latency_ms,
        throughput_mbps=throughput_mbps
    )


def run_benchmark(port: int, header_size: int, data_size: int, chunk_sizes: List[int], num_runs: int = 3) -> List[BenchmarkResult]:
    """Run benchmarks for different chunk sizes with multiple runs per size."""
    header = b"H" * header_size
    data = b"D" * data_size
    
    results = []
    
    for chunk_size in chunk_sizes:
        print(f"\nBenchmarking chunk size: {chunk_size} bytes")
        run_results = []
        
        for i in range(num_runs):
            result = run_client_test(port, header, data, chunk_size)
            run_results.append(result)
            print(f"  Run {i+1}: Latency={result.latency_ms:.2f}ms, Throughput={result.throughput_mbps:.2f}Mbps")
            time.sleep(0.5)  # Brief pause between runs
        
        # Average the results
        avg_latency = statistics.mean(r.latency_ms for r in run_results)
        avg_throughput = statistics.mean(r.throughput_mbps for r in run_results)
        
        results.append(BenchmarkResult(
            chunk_size=chunk_size,
            num_frames=run_results[0].num_frames,
            latency_ms=avg_latency,
            throughput_mbps=avg_throughput
        ))
        
    return results


def main():
    parser = argparse.ArgumentParser(description="ZeroMQ Frame Size Benchmark")
    parser.add_argument("--port", type=int, default=5555, help="ZeroMQ port")
    parser.add_argument("--header-size", type=int, default=100, help="Size of the header in bytes")
    parser.add_argument("--data-size", type=int, default=500000, help="Size of the binary data in bytes")
    parser.add_argument("--runs", type=int, default=5, help="Number of test runs per chunk size")
    args = parser.parse_args()
    
    # Test various chunk sizes
    chunk_sizes = [
        1024,       # 1KB
        8192,       # 8KB
        16384,      # 16KB
        32768,      # 32KB
        65536,      # 64KB
        131072,     # 128KB
        262144,     # 256KB
        524288,     # 512KB (single frame for data)
    ]
    
    # Start server
    server_thread = start_server(args.port)
    time.sleep(1)  # Give server time to start
    
    print(f"Starting benchmark with {args.header_size} byte header and {args.data_size/1000:.1f}KB data")
    print(f"Running {args.runs} tests per chunk size")
    
    try:
        results = run_benchmark(args.port, args.header_size, args.data_size, chunk_sizes, args.runs)
        
        # Print results table
        print("\n=== BENCHMARK RESULTS ===")
        print(f"{'Chunk Size (bytes)':<20} {'# Frames':<10} {'Avg Latency (ms)':<20} {'Avg Throughput (Mbps)':<20}")
        print("-" * 70)
        
        for result in results:
            print(f"{result.chunk_size:<20} {result.num_frames:<10} {result.latency_ms:<20.2f} {result.throughput_mbps:<20.2f}")
            
        # Find best results
        min_latency = min(results, key=lambda x: x.latency_ms)
        max_throughput = max(results, key=lambda x: x.throughput_mbps)
        
        print("\n=== SUMMARY ===")
        print(f"Lowest latency: {min_latency.chunk_size} bytes ({min_latency.latency_ms:.2f}ms)")
        print(f"Highest throughput: {max_throughput.chunk_size} bytes ({max_throughput.throughput_mbps:.2f}Mbps)")
        
    finally:
        # Clean up
        zmq.Context.instance().term()
        time.sleep(0.5)  # Give server time to shut down

if __name__ == "__main__":
    main()